File: task13/task13.cpp
#include <iostream>
#include <vector>
#include <cmath>

class CaesarCipher
{
public:
    static std::vector<std::string> movingShift(const std::string &s, int shift);
    static std::string demovingShift(std::vector<std::string> &s, int shift);
};

std::vector<std::string> CaesarCipher::movingShift(const std::string &s, int shift)
{
    std::vector<std::string> result;
    int len_s = s.length();
    std::string s_copy = s.substr(0, len_s);
    int splits = ceil(len_s / 5.0);

    for (int i = 0; i < len_s; i++)
    {
        if (isalpha(s_copy[i]))
        {
            char to_sub = (islower(s_copy[i]) ? 'a' : 'A');
            s_copy[i] = to_sub + ((s_copy[i] - to_sub) + shift) % 26;
        }
        ++shift;
    }

    for (int i = 0; i < len_s; i += splits)
    {
        result.push_back(s_copy.substr(i, splits));
    }

    int len_res = result.size();
    while (len_res < 5)
    {
        result.push_back("");
        ++len_res;
    }

    return result;
}

std::string CaesarCipher::demovingShift(std::vector<std::string> &s, int shift)
{
    std::string result;

    for (const std::string curr_s : s)
    {
        for (const char c : curr_s)
        {
            if (isalpha(c))
            {
                char to_sub = (islower(c) ? 'a' : 'A');
                char last_letter = (islower(c) ? 'z' : 'Z');
                int diff = (((c - to_sub) - shift) % 26);
                if (diff < 0)
                {
                    result += last_letter + diff + 1;
                }
                else
                {
                    result += to_sub + diff;
                }
            }
            else
            {
                result += c;
            }

            ++shift;
        }
    }

    return result;
}

int main(int argc, char const *argv[])
{
    std::string u = "I should have known that you would have a perfect answer for me!!!";

    std::vector<std::string> result = CaesarCipher::movingShift(u, 1);

    std::cout << "[ ";
    for (const std::string &s : result)
    {
        std::cout << s << ", ";
    }
    std::cout << " ]\n";

    std::cout << "\nAfter demoving: " << CaesarCipher::demovingShift(result, 1) << '\n';

    return 0;
}

File: task13/task13.py
from math import ceil


def moving_shift(s, shift):
    len_list = ceil(len(s) / 5)
    current_len = 0
    result = []

    current_s = ''
    for c in s:
        if c.isalpha():
            is_up = c.isupper()
            if is_up:
                current_s += chr(ord('A') + ((ord(c) - ord('A') + shift) % 26))
            else:
                current_s += chr(ord('a') + ((ord(c) - ord('a') + shift) % 26))
        else:
            current_s += c
        shift += 1

    while len(current_s) > 0:
        result.append(current_s[:len_list])
        current_s = current_s[len_list:]

    while len(result) < 5:
        result.append('')

    return result


def demoving_shift(s, shift):
    s = ''.join(s)
    result = ''

    for c in s:
        if c.isalpha():
            is_up = c.isupper()
            if is_up:
                result += chr(ord('A') + ((ord(c) - ord('A') - shift) % 26))
            else:
                result += chr(ord('a') + ((ord(c) - ord('a') - shift) % 26))
        else:
            result += c
        shift += 1

    return result


def main():
    u = "I should have known that you would have a perfect answer for me!!!"
    v = ["J vltasl rlhr ", "zdfog odxr ypw",
         " atasl rlhr p ", "gwkzzyq zntyhv", " lvz wp!!!"]

    result = moving_shift(u, 1)
    print(result)
    print(result == v)


if __name__ == '__main__':
    main()
